br/pkg/lightning/importer/dup_detect_test.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package importer
+
+import (
+	"context"
+	"testing"
+
+	"github.com/pingcap/tidb/br/pkg/lightning/duplicate"
+	"github.com/pingcap/tidb/util/extsort"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/exp/slices"
+)
+
+func TestErrorOnDup(t *testing.T) {
+	h := &errorOnDup{}
+	err := h.Begin([]byte("key"))
+	require.Error(t, err)
+	require.NoError(t, h.Close())
+}
+
+func TestReplaceOnDup(t *testing.T) {
+	runDupHandlerTest(t,
+		func(w extsort.Writer) duplicate.Handler { return &replaceOnDup{w: w} },
+		[]dupRecord{{
+			[]byte("key1"), [][]byte{[]byte("01"), []byte("02"), []byte("03")}},
+			{[]byte("key2"), [][]byte{[]byte("11"), []byte("12"), []byte("13")}}},
+		[][]byte{[]byte("01"), []byte("02"), []byte("11"), []byte("12")},
```

Comment:
seems the caller can not and does not need to know which key the duplicated keyID belongs to 🤔

for multiple row conflicts on different PK/UK, can we prove that exactly one keyID is always not written to extsort? Maybe add a unit test to explain it.

Reply:
Now I understand that the "duplicate-with" is transitive for keyID, if keyID `a` is duplicated with `b` on index `idx1`, `b` is duplicated with `c` on `idx2`, then `a` is also duplicated with `c` even though these two lines are not duplicated on any index 🤔 it's important to address this behavour in document.

Reply:
Good catch! I will think about how to solve this problem. Do you have any idea?

Reply:
I found it's not easy to make `ignore` behavior consistent with `IGNORE INTO`. 
We can either document the difference with `IGNORE INTO` or do not support `ignore` at all.
Do you think which one is better?

Reply:
I think we can document it, and maybe give current implementation other name than `IGNORE` in case we need exactly IGNORE behaviour in future. And I checked with chatGPT IGNORE requires O(N*avg(conflict group size)) algorithm.

BTW preprocess duplicate detection is done before dropping index?

Reply:
> BTW preprocess duplicate detection is done before dropping index?

Yes, this is to ensure adding unique index can always succeed.